src/agent.py

```diff
@@ -68,31 +83,89 @@ def _on_agent_relation_joined(self, event: ops.RelationJoinedEvent) -> None:
             return
 
         logger.info("%s relation joined.", event.relation.name)
-        self.charm.unit.status = ops.MaintenanceStatus("Setting up relation.")
+        self.charm.unit.status = ops.MaintenanceStatus(
+            f"Setting up '{event.relation.name}' relation."
+        )
+
+        relation_data = self.state.agent_meta.get_jenkins_agent_v0_interface_dict()
+        logger.debug("Agent relation data set: %s", relation_data)
+        event.relation.data[self.charm.unit].update(relation_data)
+
+    def _on_slave_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        """Handle slave relation changed event.
+
+        Args:
+            event: The event fired when slave relation data has changed.
+        """
+        logger.info("%s relation changed.", event.relation.name)
+
+        if self.state.jenkins_config or self.state.agent_relation_credentials:
+            logger.warning(
+                "Jenkins configuration already exists. Ignoring %s relation.", event.relation.name
+            )
+            return
+
+        container = self.charm.unit.get_container(self.state.jenkins_agent_service_name)
+        if not container.can_connect():
+            logger.warning("Jenkins agent container not yet ready. Deferring.")
+            event.defer()
+            return
 
-        # handle slave relation for backwards compatibility.
-        if event.relation.name == SLAVE_RELATION:
-            logger.info(
-                "Slave relation data set: %s",
-                self.state.agent_meta.get_jenkins_slave_interface_dict(),
+        # Check if the pebble service has started and set agent ready.
+        if container.exists(str(server.AGENT_READY_PATH)):
+            logger.warning("Given agent already registered. Skipping.")
+            return
```

Comment:
It does indeed, however, the checks are run periodically in second scale, but in this case we need it to be a bit more faster than what pebble tells us because the already locked-in agent trying to register with another name-token credentials could cause problems (it has chance to take up additional availability).

Reply:
Yes indeed, I checked pebble code even calling the check endpoint doesn't refresh the status, so indeed information may not be fresh.